### PR TITLE
feat: OSS local cache layer

### LINKED ISSUE #28: Cache file in local file system
# Target
Provides a local file system cache layer over object store

# Design

Add an `ObjectStore` implementor `CachedStore`, which is made up of:
- Local Store: a local file system cache
- Remote Store: a remote OSS service
- LRU Manager: a cache eviction policy (LRU) to keep the local store size under control.

## Read

On serving a read request, the `CachedStore` will first check if the object is in the local store. If not, it will fetch it from the remote store and store it in the local store. To simplify the implementation, all the read sources returned are based on `LocalStore`. Ascii art below:

Workflow
```
(read)
       Serve I/O requests
              ▲ │
            4 │ │1
              │ │
   ┌──────────┴─▼─────────┐
   │                      │
   │  Object Store Cache  │
   │                      │
   └─┬─▲────────────┬─▲───┘
  2.1│ │3.1      2.2│ │3.2
     │ │            │ │
 ┌───▼─┴──┐      ┌──▼─┴─────┐
 │Local FS│      │Remote OSS│
 └────────┘      └──────────┘
```

## Write

For write requests, we will write the content to both underlying stores.

## Restart

To suit some deploy scenarios that aren't stateless, `ObjectStore` will try to load all existing entries from `LocalStore` to `LRU Manager`.

## Purge

Both read and write operations may trigger purge on `LocalStore`. The purge policy is defined by the `LRU Manager`. It will get a delete list from `LRU Manager` and delete the list from `LocalStore`.

To ensure the total size of `LocalStore` is always less than the threshold, `CachedStore` will first purge enough space for the incoming new objects.

# Other features

## Config
Except for the basic configurations like the underlying provider of `LocalStore` and `RemoteStore`, threshold of `LocalStore` etc., we can also configure other behaviors like whether to cache write request, how to act on restart or cache evict policy.

## Async Cache
It is doable to feed the data from `RemoteStore` directly to the user and write `LocalStore` asynchronously.

### PR DESCRIPTION
**Pull request by waynexia**
# Which issue does this PR close?

Closes #28 

# Rationale for this change
 

See #28 
# What changes are included in this PR?

A wrapper `CachedStore` over two `ObjectStore` instances that provides cache utility.

# Are there any user-facing changes?

No

# How does this change test

New unit tests of `CachedStore`

### CODE REVIEW

File: components/object_store/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        self.total_size += size;
+        return (removed, Some(EntryGuard { size }));
+    }
+
+    fn remove_guard(&mut self, guard: EntryGuard) {
+        self.total_size -= guard.size;
+    }
+
+    fn consume_guard(&mut self, guard: EntryGuard, location: &Path) {
+        let prev_cache = self.cached_entries.push(
```
[the remark this answers is not in the document]
Comment (waynexia) [reply components/object_store/src/cache.rs:214]: `push()` returns an Option. I have changed the following `if let` to this form.

File: components/object_store/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    #[must_use]
+    fn reserve(&mut self, size: usize) -> (Vec<Path>, Option<EntryGuard>) {
+        if self.total_size + size <= self.max_size {
+            self.total_size += size;
+            return (Vec::new(), Some(EntryGuard { size }));
+        }
+
+        let mut removed = Vec::new();
+        while self.total_size + size > self.max_size {
+            let popped = self.cached_entries.pop_lru();
```
[the remark this answers is not in the document]
Comment (waynexia) [reply components/object_store/src/cache.rs:197]: Changed the following `if let`

File: components/object_store/src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    /// be treat as a cache read operation.
+    fn contains(&mut self, location: &Path) -> bool {
+        self.cached_entries.get(&location.to_string()).is_some()
+    }
+
+    /// Reserve space for a new entry. Returns a guard that stands for the
+    /// space. It needs to be consumed by either [remove_guard] (cancel this
+    /// reserve operation) or [consume_guard] (confirm this reserve operation).
+    #[must_use]
+    fn reserve(&mut self, size: usize) -> (Vec<Path>, Option<EntryGuard>) {
```
[the remark this answers is not in the document]
Comment (waynexia) [reply components/object_store/src/cache.rs:189]: 1. the first element `Vec` stands for those paths that need to be removed. The `EntryGuard` (has changed to `ReserveResult`) is optional. When it's `None` means this reserve operation is failed. They are two separate results and I think it's better to return a tuple.
2. Done.

File: components/object_store/src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    /// Try putting object to local store. If local store cannot make enough
+    /// space for the object, this function will skip putting it and return Ok.
+    async fn put_local(&self, location: &Path, bytes: Bytes) -> Result<()> {
+        let required_size = bytes.len();
+
+        let mut state = self.state.lock().await;
+        let (removed, guard) = state.reserve(required_size);
+        Self::remove_paths(self.local_store.as_ref(), removed).await?;
```
[the remark this answers is not in the document]
Comment (waynexia) [reply components/object_store/src/cache.rs:53]: done

File: components/object_store/src/cache.rs [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+        } else {
+            let remote_obj = self.remote_store.get(location).await?;
+            let bytes = remote_obj.bytes().await?;
+            let _ = self.put_local(location, bytes).await?;
+            self.local_store.get_range(location, range).await
+        }
+    }
+
+    async fn head(&self, location: &Path) -> Result<ObjectMeta> {
+        self.remote_store.head(location).await
```
[the remark this answers is not in the document]
Comment (waynexia) [reply components/object_store/src/cache.rs:115]: I left a TODO for this

File: components/object_store/Cargo.toml [hunk 6]
```diff
@@ -11,3 +11,9 @@ futures = "0.3"
 upstream = { package = "object_store", version =  "0.1.0" }
 oss-rust-sdk = "0.4.0"
 snafu = { version = "0.6.10", features = ["backtraces"] }
+lru = "0.7.6"
+chrono = "0.4.19"
+
+[dev-dependencies]
+tempfile = "3.1.0"
+tokio = { version = "1.0", features = ["full"] }
```
Comment (ShiKaiWi) [inline components/object_store/Cargo.toml:19]: Add a newline here?

**Review by ShiKaiWi — APPROVED** [review]
👍👍

**Pull request merged**